includes/class-amp-service-workers.php

```diff
@@ -0,0 +1,146 @@
+<?php
+/**
+ * AMP Service Workers.
+ *
+ * @package AMP
+ * @since 1.0
+ */
+
+/**
+ * Class AMP_Service_Workers.
+ */
+class AMP_Service_Workers {
+
+	/**
+	 * Query var that is used to signal a request to install the service worker in an iframe.
+	 *
+	 * @link https://www.ampproject.org/docs/reference/components/amp-install-serviceworker#data-iframe-src-(optional)
+	 */
+	const INSTALL_SERVICE_WORKER_IFRAME_QUERY_VAR = 'amp_install_service_worker_iframe';
+
+	/**
+	 * Init.
+	 */
+	public static function init() {
+		if ( ! function_exists( 'wp_register_service_worker' ) ) {
+			return;
+		}
+
+		add_filter( 'query_vars', function( $vars ) {
+			$vars[] = self::INSTALL_SERVICE_WORKER_IFRAME_QUERY_VAR;
+			return $vars;
+		} );
+
+		wp_register_service_worker(
+			'amp-asset-caching',
+			amp_get_asset_url( 'js/amp-service-worker-asset-cache.js' )
+		);
+		add_action( 'wp_default_service_workers', array( __CLASS__, 'register_script' ) );
+		add_action( 'parse_request', array( __CLASS__, 'handle_service_worker_iframe_install' ) );
+		add_action( 'wp', array( __CLASS__, 'add_install_hooks' ) );
+	}
+
+	/**
+	 * Add hooks to install the service worker from AMP page.
+	 */
+	public static function add_install_hooks() {
+		if ( current_theme_supports( 'amp' ) && is_amp_endpoint() ) {
+			add_action( 'wp_footer', array( __CLASS__, 'install_service_worker' ) );
+
+			// Prevent validation error due to the script that installs the service worker on non-AMP pages.
+			$priority = has_action( 'wp_print_scripts', 'wp_print_service_workers' );
+			if ( false !== $priority ) {
+				remove_action( 'wp_print_scripts', 'wp_print_service_workers', $priority );
+			}
+		}
+		add_action( 'amp_post_template_footer', array( __CLASS__, 'install_service_worker' ) );
+	}
+
+	/**
+	 * Install service worker(s).
+	 *
+	 * @since 1.0
+	 * @see wp_print_service_workers()
+	 * @link https://github.com/xwp/pwa-wp
+	 */
+	public static function install_service_worker() {
+		if ( ! function_exists( 'wp_service_workers' ) || ! function_exists( 'wp_get_service_worker_url' ) ) {
+			return;
+		}
+
+		// Get the frontend-scoped service worker scripts.
+		$front_handles = array();
+		foreach ( wp_service_workers()->registered as $handle => $item ) {
+			if ( 'all' === $item->args['scope'] || 'front' === $item->args['scope'] ) {
```

Comment:
This needs to be updated to use the `WP_Service_Worker::SCOPE_FRONT` constant, in a bitwise `&` condition.